// @civic/auth/dist/vanillajs/auth/types/AuthTypes.js
/**
 * Error codes for CivicAuth errors
 */
export var CivicAuthErrorCode;
(function (CivicAuthErrorCode) {
    CivicAuthErrorCode["CONFIG_REQUIRED"] = "CONFIG_REQUIRED";
    CivicAuthErrorCode["INIT_FAILED"] = "INIT_FAILED";
    CivicAuthErrorCode["ENDPOINTS_NOT_INITIALIZED"] = "ENDPOINTS_NOT_INITIALIZED";
    CivicAuthErrorCode["CONTAINER_NOT_FOUND"] = "CONTAINER_NOT_FOUND";
    CivicAuthErrorCode["AUTH_PROCESS_TIMEOUT"] = "AUTH_PROCESS_TIMEOUT";
    CivicAuthErrorCode["IFRAME_LOAD_ERROR"] = "IFRAME_LOAD_ERROR";
    CivicAuthErrorCode["INVALID_MESSAGE"] = "INVALID_MESSAGE";
    CivicAuthErrorCode["LOGOUT_FAILED"] = "LOGOUT_FAILED";
    CivicAuthErrorCode["POPUP_BLOCKED"] = "popup_blocked";
    CivicAuthErrorCode["USER_CANCELLED"] = "user_cancelled";
    CivicAuthErrorCode["CONFIGURATION_ERROR"] = "configuration_error";
    CivicAuthErrorCode["TOKEN_REFRESH_FAILED"] = "token_refresh_failed";
    CivicAuthErrorCode["SESSION_NOT_FOUND"] = "session_not_found";
    CivicAuthErrorCode["STORAGE_ERROR"] = "storage_error";
    CivicAuthErrorCode["IFRAME_NOT_FOUND"] = "iframe_not_found";
    CivicAuthErrorCode["INTERNAL_ERROR"] = "internal_error";
})(CivicAuthErrorCode || (CivicAuthErrorCode = {}));
/**
 * Constants for the auth client
 */
export const CIVIC_AUTH_CONSTANTS = {
    DEFAULT_IFRAME_ID: "civic-auth-iframe",
    DEFAULT_AUTH_PROCESS_TIMEOUT: 60000, // 60 seconds
    SUCCESS_SIGNAL_ID: "civic-auth-success-signal",
    ERROR_SIGNAL_ID: "civic-auth-error-signal",
};
export class CivicAuthError extends Error {
    code;
    constructor(message, code) {
        super(message);
        this.code = code;
        this.name = "CivicAuthError";
    }
}
